### PR TITLE
bpo-29960 _random.Random state corrupted on exception.

### PR DESCRIPTION
**Pull request by bladebryan**
Changes the _random.Random.setstate() function so that if it raises an exception the state of the generator is unchanged.

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://cpython-devguide.readthedocs.io/pullrequest.html#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

File: Modules/_randommodule.c [hunk 1]
```diff
@@ -375,6 +376,8 @@ random_setstate(RandomObject *self, PyObject *state)
         return NULL;
     }
     self->index = (int)index;
+    for (i=0; i<N; i++)
```
Comment (serhiy-storchaka) [inline Modules/_randommodule.c]: Add spaces around operators `=` and `<`.

**Review by serhiy-storchaka — COMMENTED** [review]
Please add an entry in `Misc/NEWS`.

File: Misc/NEWS [hunk 2]
```diff
@@ -10,6 +10,9 @@ What's New in Python 3.7.0 alpha 1?
 Core and Builtins
 -----------------
 
+- bpo-29960: Preserve generator state when _random.Random.setstate()
+  raises an exception.
```
Comment (serhiy-storchaka) [inline Misc/NEWS]: Add "Patch by $yourname." Use two spaces between sentences.

**Comment by rhettinger** [comment]
The 14 files changed seems wrong.  I think a lot of other stuff may have inadvertently been included.

**Comment by bladebryan** [comment]
I must have done something wrong in rebase. This pull request has gotten too messy, so I'll close it and create another.

**Pull request closed without merging**